miden/src/tools/mod.rs

```diff
@@ -21,31 +20,67 @@ impl Analyze {
         let program_info: ProgramInfo =
             analyze(program.as_str(), program_inputs).expect("Could not retreive program info");
 
-        let total_vm_cycles = program_info.total_vm_cycles();
-        let total_noops = program_info.total_noops();
-
-        println!("Total Number of VM Cycles: {}", total_vm_cycles);
-        println!("Total Number of NOOPs executed: {}", total_noops);
-
+        print_stats(&program_info);
         Ok(())
     }
 }
 
+/// Utility function to print the following stats about the program being run
+/// - Total Number of VM Cycles
+/// - Total Number of NOOPs executed:
+/// - Number of cycles per assembly instruction
+fn print_stats(program_info: &ProgramInfo) {
+    let total_vm_cycles = program_info.total_vm_cycles();
+    let total_noops = program_info.total_noops();
+    let asmop_info_list = program_info.asmop_info_list();
+    println!("Total Number of VM Cycles: {}\n", total_vm_cycles);
+    println!("Total Number of NOOPs executed: {}\n", total_noops);
+    println!(
+        "{0: <20} | {1: <20} | {2: <20} | {3: <20}",
+        "AsmOp", "Instruction Cycles", "Frequency", "Total Cycles"
+    );
+    println!("{:-<1$}", "", 80);
+    for op_info in asmop_info_list {
+        println!(
+            "{0: <20} | {1: <20} | {2: <20} | {3: <20}",
+            op_info.0,
+            op_info.1,
+            op_info.2,
+            op_info.2 * op_info.1 as usize
+        );
+    }
+}
+
 /// Contains info of a program. Used for program analysis. Contains the following fields:
 /// - total_vm_cycles (vm cycles it takes to execute the entire script)
 /// - total_noops (total noops executed as part of a program)
+/// - asmop_info_list (contains a list of assembly instructions and  the number of vm cycles it
+/// takes to execute the instruction and the number of times the instruction is run as part of the
+/// given program.)
 #[derive(Debug, Eq, PartialEq)]
 pub struct ProgramInfo {
     total_vm_cycles: usize,
     total_noops: usize,
+    asmop_info_list: Vec<(String, u8, usize)>,
 }
 
 impl ProgramInfo {
-    /// Creates a new ProgramInfo object
-    pub fn new(total_vm_cycles: usize, total_noops: usize) -> ProgramInfo {
+    /// Creates a new ProgramInfo object from the specified `total_vm_cycles`, `total_noops` and
+    /// `asmop_info_list`.
+    /// * total_vm_cycles: Total number of VM cycles required to run the given program.
+    /// * total_noops: Total number of NOOPs executed as part of the program.
+    /// * asmop_info_list: Vector of tuples with a list of assembly instructions and the number of
+    /// vm cycles it takes to execute the instruction and the number of times the instruction is
+    /// run as part of the given program.
+    pub fn new(
+        total_vm_cycles: usize,
+        total_noops: usize,
+        asmop_info_list: Vec<(String, u8, usize)>,
```

Comment:
Maybe we should make `(String, u8, usize)` into its one struct?

Reply:
Done. Does `AsmOpDebugInfo` sound right as the name of the struct?